libqfieldsync/offliners.py

```diff
@@ -0,0 +1,401 @@
+            for i in range(new_layer.fields().count()):
+                type = new_layer.fields().at(i).type()
+                if type == QVariant.StringList or type == QVariant.List:
+                    attrs[i] = QgsJsonUtils.encodeValue(attrs[i])
+
+            feature.setFields(new_fields)
+            feature.setAttributes(attrs)
+            new_layer.addFeature(feature)
+
+    return new_layer.source()
+
+
+def update_data_provider(layer: QgsVectorLayer, source: str) -> None:
+    # Mark as offline layer
+    layer.setCustomProperty(CUSTOM_PROPERTY_IS_OFFLINE_EDITABLE, True)
+
+    # store original layer source and information
+    layer.setCustomProperty(CUSTOM_PROPERTY_REMOTE_SOURCE, layer.source())
+    layer.setCustomProperty(CUSTOM_PROPERTY_REMOTE_PROVIDER, layer.providerType())
+    layer.setCustomProperty(CUSTOM_PROPERTY_ORIGINAL_LAYERID, layer.id())
+    layer.setCustomProperty(CUSTOM_PROPERTY_LAYERNAME_SUFFIX, "")
+
+    # TODO: review this code, why do we connect defaultvalueclause to not null constraints?
+    #
+    # remove constrainst of fields that use defaultValueClauses from provider on original
+    not_null_field_names = []
+    source_fields = layer.fields()
+    for field in source_fields:
+        if layer.dataProvider().defaultValueClause(
+            layer.fields().fieldOriginIndex(layer.fields().indexOf(field.name()))
+        ):
+            not_null_field_names.append(field.name())
+
+    layer.setDataSource(source, layer.name(), "ogr")
+
+    for field in source_fields:
+        index = layer.fields().indexOf(field.name())
+        if index > -1:
+            # restore unique value constraints coming from original data provider
+            if field.constraints().constraints() & QgsFieldConstraints.ConstraintUnique:
+                layer.setFieldConstraint(index, QgsFieldConstraints.ConstraintUnique)
+
+            # remove any undesired not null constraints coming from original data provider
+            if field.name() in not_null_field_names:
+                layer.removeFieldConstraint(
+                    index, QgsFieldConstraints.ConstraintNotNull
+                )
+
+
+def convert_to_offline_project(
+    offline_gpkg_path: str,
+    offline_layers: Optional[List[QgsMapLayer]],
+    bbox: QgsRectangle = QgsRectangle(),
+) -> None:
+    """Converts the currently loaded QgsProject to an offline project.
+    Offline layers are written to ``offline_gpkg_path``. Only valid vector layers are written.
+    If ``layer_ids`` is specified, only layers present in this list are written.
+    If ``bbox`` is specified, only features within this ``bbox`` are written.
+    """
+    project = QgsProject.instance()
+
+    driver = ogr.GetDriverByName("GPKG")
+    data_source = driver.CreateDataSource(offline_gpkg_path)
+
+    # QgsOfflineEditing sets PRAGMA FOREIGN_KEY -- is this required?
+    # QgsOfflineEditing calls Initialize Spatial Metadata -- is this required? Probably a spatialite leftover
```

Comment:
```suggestion
```

IMO we don't want to necessarily deal with a at every point in time completely consistent offline dataset. If we have foreign key issues on a qfield log, that's a PITA for user, project manager and developer. Better deal with it a synchronization level.

Reply:
Agreed. Once we have the data (or the delta) we can manage it from there.

On the comments, I would prefer to keep them and highligh the differences with QGIS implementation, so future us does not ask questions.